### PR TITLE
Enforce maximum outstanding publisher confirms, if set

### PR DESCRIPTION
**Pull request by lukebakken**
Part of #1682

### CODE REVIEW

File: projects/RabbitMQ.Client/Impl/Channel.PublisherConfirms.cs [hunk 1]
```diff
@@ -329,5 +342,49 @@ await publisherConfirmationInfo.MaybeWaitForConfirmationAsync(cancellationToken)
...
+
+            return default;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void MaybeBlockPublishers()
+        {
+            if (_publisherConfirmationTrackingEnabled)
+            {
+                if (_maxOutstandingPublisherConfirmations is not null
```
Comment (danielmarbach) [inline projects/RabbitMQ.Client/Impl/Channel.PublisherConfirms.cs:369]: Wouldn't it be way simpler to have another semaphore slim that is set to the number of outstanding confirms allowed or not set in the unlimited mode?

Comment (lukebakken) [reply projects/RabbitMQ.Client/Impl/Channel.PublisherConfirms.cs:369]: That makes _WAY MORE SENSE_ 🤦‍♂️

File: projects/RabbitMQ.Client/Impl/Channel.PublisherConfirms.cs [hunk 2]
```diff
@@ -270,6 +282,18 @@ await _confirmSemaphore.WaitAsync(reason.CancellationToken)
         {
             if (_publisherConfirmationsEnabled)
             {
+                if (_publisherConfirmationTrackingEnabled)
+                {
+                    if (_maxOutstandingPublisherConfirmations is not null)
+                    {
+                        int percentOfMax = _confirmsTaskCompletionSources.Count / (int)_maxOutstandingPublisherConfirmations;
+                        if (percentOfMax > 0.5)
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply projects/RabbitMQ.Client/Impl/Channel.PublisherConfirms.cs:290]: Here is a spike https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1704

**Comment by danielmarbach** [comment]
I have looked at the client code changes that have been recently added in addition to those changes and overall the inlining of the channel base and the partial split helps to better navigate the complexity of the code. 

My biggest concern after reading through the changes is that all the flag checking and the hidden acquiring and releasing of the confirmation semaphore plus the manipulation of the sequence number in methods that assume the semaphore is held makes it non-trivial to understand the flow. I think this structure could quickly become a maintenance burden overtime and be the source of subtle bugs.

I could not spend more time digging into the code because the time I had available was eaten up by providing the spike. Once those discussions settle, I might find some time to think how the code can be restructured.

File: projects/RabbitMQ.Client/Impl/Channel.PublisherConfirms.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukebakken) [reply projects/RabbitMQ.Client/Impl/Channel.PublisherConfirms.cs:290]: Cool! Thanks! The code I have is really just to get the discussion going. I'll look at your PR tomorrow.

**Comment by lukebakken** [comment]
> My biggest concern after reading through the changes is that all the flag checking and the hidden acquiring and releasing of the confirmation semaphore plus the manipulation of the sequence number in methods that assume the semaphore is held makes it non-trivial to understand the flow.

I agree. I will add a lot more comments to the code. At this point, unless a bug or improvement is found, I'd like to keep it as is because it's all "behind the scenes" anyway and we can change it in the future.

**Pull request merged**